Expose cookie Partition Key in Web Inspector

sysrqb opened the pull request:
#### a7fa91e9eceb1305b7b26feb1e3d55bc69b57486
<pre>
Expose cookie Partition Key in Web Inspector
<a href="https://bugs.webkit.org/show_bug.cgi?id=279980">https://bugs.webkit.org/show_bug.cgi?id=279980</a>
<a href="https://rdar.apple.com/136293236">rdar://136293236</a>

Reviewed by BJ Burg and Devin Rousso.

This patch adds a new column in the cookie storage view that displays a
cookie&apos;s partition key. Currently, the column is hidden by default. The
partition key is also shown and can also be set in the popover. However, the
values are not currently plumbed into the underlying cookie stores, so this
change doesn&apos;t have any affect on the underlying cookie.

* Source/JavaScriptCore/inspector/protocol/Page.json:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::buildObjectForCookie):
(WebCore::parseCookieObject):
(WebCore::InspectorPageAgent::setCookie):
* Source/WebCore/inspector/agents/InspectorPageAgent.h:
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::setRawCookie):
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setCookie):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/ShouldPartitionCookie.h: Added.
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/Cookie.js:
(WI.Cookie.parseSetCookieResponseHeader):
(WI.Cookie.prototype.get partitionKey):
(WI.Cookie.prototype.get partitioned):
(WI.Cookie.prototype.equals):
(WI.Cookie.prototype.toProtocol):
(WI.Cookie):
* Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
(WI.CookiePopover):
(WI.CookiePopover.prototype.get serializedData):
(WI.CookiePopover.prototype.show):
* Source/WebInspectorUI/UserInterface/Views/CookieStorageContentView.js:
(WI.CookieStorageContentView.prototype.initialLayout):
(WI.CookieStorageContentView.prototype._generateSortComparator):
(WI.CookieStorageContentView.prototype._showCookiePopover):
(WI.CookieStorageContentView.prototype.async _willDismissCookiePopover):
(WI.CookieStorageContentView.prototype._formatCookiePropertyForColumn):
(WI.CookieStorageContentView):
* Source/WebInspectorUI/UserInterface/Views/ResourceCookiesContentView.js:
(WI.ResourceCookiesContentView.prototype.tablePopulateCell):
(WI.ResourceCookiesContentView.prototype._refreshResponseCookiesSection):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setRawCookie):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::setRawCookie):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: <a href="https://commits.webkit.org/288103@main">https://commits.webkit.org/288103@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4775866938aca3b4229f54bea82e34e91d7c75b7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81871 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1397 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35825 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86425 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32867 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83977 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1430 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9220 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63854 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21583 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1062 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74506 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44140 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/957 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28682 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31320 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74862 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29292 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87855 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80925 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9111 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6493 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72220 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9296 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70327 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71444 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15531 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14454 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103337 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12690 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9062 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14594 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25090 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8903 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12426 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -119,6 +119,7 @@
                 { "name": "value", "type": "string", "description": "Cookie value." },
                 { "name": "domain", "type": "string", "description": "Cookie domain." },
                 { "name": "path", "type": "string", "description": "Cookie path." },
+                { "name": "partitionKey", "type": "string", "description": "Cookie partition key." },
```
I think this needs to be an optional property, since older versions of iOS that we inspect won't support it.

dcrousso replied:
actually it's totally fine for it to be non-optional here as older iOS versions have their own frozen protocol files that Web Inspector will load from when it first connects.  this file in it's current state will only be used for things going forward

burg commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -143,6 +150,7 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
             data.value = "";
             data.domain = urlComponents.host;
             data.path = urlComponents.path;
+            data.partitionKey = null;
```
This is unnecessary.

burg commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -211,6 +219,9 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
         this._pathInputElement = createInputRow("path", WI.unlocalizedString("Path"), "text", data.path).inputElement;
 
+        if (options.includePartitionKey)
+            this._partitionKeyInputElement = createInputRow("partitionKey", WI.unlocalizedString("Partition Key"), "text", data.partitionKey).inputElement;
```
This should be a localized string because unlike the other columns, it is not a literal that you'd find in the cookie itself.  In CHIPS it would be `Partitioned`.

burg commented on Source/WebInspectorUI/UserInterface/Views/CookieStorageContentView.js:
```diff
@@ -271,6 +271,13 @@ WI.CookieStorageContentView = class CookieStorageContentView extends WI.ContentV
             initialWidth: 100,
         });
 
+        this._partitionKeyColumn = new WI.TableColumn("partitionKey", WI.unlocalizedString("Partition Key"), {
```
Ditto the above comment on localization

burg commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -63,6 +64,10 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
             return null;
 
+        let partitionKey = this._partitionKeyInputElement.value || this._partitionKeyInputElement.placeholder;
+        if (partitionKey)
```
Looks like a typo, this should be `if (!partitionKey)`

burg commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -585,11 +585,13 @@ static Inspector::Protocol::Page::CookieSameSitePolicy cookieSameSitePolicyJSON(
 
 static Ref<Inspector::Protocol::Page::Cookie> buildObjectForCookie(const Cookie& cookie)
 {
+    // FIXME: Add partitionKey
```
Please fill in a bugzilla or radar link.

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -604,6 +605,8 @@ static Ref<Inspector::Protocol::Page::Cookie> buildObjectForCookie(const Cookie&
         .setSecure(cookie.secure)
         .setSameSite(cookieSameSitePolicyJSON(cookie.sameSite))
         .release();
+    inspectorCookie->setPartitionKey(cookie.partitionKey);
```
will the `partitionKey` always exist?  or is it sometimes `""`/`null`?  if the latter, id suggest we avoid setting it unless it's actually valid

sysrqb replied:
It is sometimes null.

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -653,69 +656,76 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::Page::
+static std::pair<std::optional<Cookie>, IsPartitionedCookie> parseCookieObject(Inspector::Protocol::ErrorString& errorString, Ref<JSON::Object>&& cookieObject)
 {
     Cookie cookie;
+    IsPartitionedCookie isPartitioned = IsPartitionedCookie::No;
```
NIT: we should move this down lower to where it's actually used

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -653,69 +656,76 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::Page::
 }
 
-static std::optional<Cookie> parseCookieObject(Inspector::Protocol::ErrorString& errorString, Ref<JSON::Object>&& cookieObject)
+static std::pair<std::optional<Cookie>, IsPartitionedCookie> parseCookieObject(Inspector::Protocol::ErrorString& errorString, Ref<JSON::Object>&& cookieObject)
```
could we reverse the order of this to be `std::optional<std::pair<Cookie, IsPartitionedCookie>>` so that we dont need `IsPartitionedCookie` if we're providing `std::nullopt`?

dcrousso commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -119,6 +120,15 @@ bool NetworkStorageSession::shouldBlockThirdPartyCookiesButKeepFirstPartyCookies
+void NetworkStorageSession::setCookie(const URL& firstParty, const Cookie& cookie, IsPartitionedCookie isPartitionedCookie)
+{
+    if (!isOptInCookiePartitioningEnabled() || isPartitionedCookie != IsPartitionedCookie::Yes)
+        setCookie(cookie);
```
should there be a `return` after this?  or is the idea that if partitioning is disabled or the specific cookie is not partitioned then we'd set it both generally and in the partition?

sysrqb replied:
Good catch, that's a bug.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Cookie.js:
```diff
@@ -155,6 +159,9 @@ WI.Cookie = class Cookie
         let secure = false;
         let httpOnly = false;
+        // partition key is implicit and is the first party site if the Partitioned attribute is present.
+        const partitionKey = null;
```
rather than have it be `null` i feel like we should set it equal to the first party site instead of having this be implicit

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Cookie.js:
```diff
@@ -301,6 +316,8 @@ WI.Cookie = class Cookie
             session: this._session,
             httpOnly: !!this._httpOnly,
+            partitionKey: this._partitionKey,
+            partitioned: this._partitioned,
```
NIT: `!!this._partitioned` to match the other booleans

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookieStorageContentView.js:
```diff
@@ -611,6 +620,8 @@ WI.CookieStorageContentView = class CookieStorageContentView extends WI.ContentV
             return cookie.domain || missingValue;
         case "path":
             return cookie.path || missingValue;
+        case "partitionKey":
+            return cookie.partitionKey || missingValue;
```
why do we need both `partitioned` and `partitionKey` if all we're showing in the UI is just the `partitionKey`?  is it not valid to assume that a cookie is `partitioned` if it's `partitionKey` is non-empty/non-null?  if so, should we have an additional column "Partitioned" that contains a checkmark if the cookie is `partitioned` regardless of it's `partitionKey`?

sysrqb replied:
CHIPS introduces a new Cookie attribute that is a boolean `Partitioned`, when that attribute is present we compute the appropriate `partitionKey` which is the site (schemeful registrable domain) of the top frame. I prefer avoiding computing the partition key in Web Inspector for the Cookie Popover, and just rely on the logic for computing it when we create the cookie. At that point, we can read the partition key from the cookie and show it in the Storage tab. This is why I added the Partitioned checkbox in the popover, and the partition key in the storage tab.

dcrousso replied:
is there no situation where the developer might want to set a custom `partitionKey`?

as far as computing the `partitionKey`, unless Web Inspector adds support for creating cookies in subframes it'll always add it to the main document, so won't it just be the  main domain?

i think also part of the oddity for me is that the cookie popover is supposed to allow for editing _all_ the details of the cookie, so not having the `partitionKey` at least visible there (i.e. in a `<input readonly>` or something) is odd

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -33,10 +33,12 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
         this._valueInputElement = null;
         this._domainInputElement = null;
         this._pathInputElement = null;
+        this._partitionKeyInputElement = null;
```
where is this used?  ideally we'd at least show the `partitionKey` (i.e. a disabled input) in the cookie popover so that developers can see it even if they can't edit it

sysrqb replied:
That is stale and was leftover. I'll remove it.

dcrousso reviewed (changes requested):
overall the logic of this looks great, but im slightly confused as to where and when we show what data, as well as whether all the pieces of info you've added are really needed

also it would be really helpful if you could attach some screenshots of what these changes look like, both in the Storage Tab and a cookie popover :)

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -598,11 +599,13 @@ static Ref<Inspector::Protocol::Page::Cookie> buildObjectForCookie(const Cookie&
         .setValue(cookie.value)
         .setDomain(cookie.domain)
         .setPath(cookie.path)
+        .setPartitionKey(cookie.partitionKey)
```
so this is actually the opposite of what I was getting at in my previous comment 😅

if the `paritionKey` is sometimes `null` then it _should_ be optional and _not_ be set unless it's valid
```cpp
if (!cookie.partitionKey.isEmpty())
    protocolCookie->setPartitionKey(cookie.partitionKey);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Cookie.js:
```diff
@@ -205,6 +210,10 @@ WI.Cookie = class Cookie
                 break;
+            case "partitioned":
+                console.assert(!attributeValue);
+                partitioned = true;
```
btw we should probably add a column inside `WI.ResourceCookiesContentView.prototype._refreshResponseCookiesSection` for "Partitioned" since this is parsing from `Set-Cookie` headers

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -119,10 +119,12 @@
                 { "name": "session", "type": "boolean", "description": "True in case of session cookie." },
                 { "name": "httpOnly", "type": "boolean", "description": "True if cookie is http-only." },
                 { "name": "secure", "type": "boolean", "description": "True if cookie is secure." },
+                { "name": "partitioned", "type": "boolean", "description": "True if cookie is partitioned." },
```
NIT: i'd prefer if this was put right before `partitionKey` (or that is put right after this) so that everything related to partitioning is in the same place

it also might be nice to add in the `description` that they're mutually exclusive, since we can assume `partitioned` is `true` if `partitionKey` is provided, and if just `partitioned` is provided without `partitionKey` then that's Web Inspector asking for the `partitionKey` to be generated

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Page.json:
```diff
@@ -123,6 +123,8 @@
                 { "name": "httpOnly", "type": "boolean", "description": "True if cookie is http-only." },
                 { "name": "secure", "type": "boolean", "description": "True if cookie is secure." },
+                { "name": "partitioned", "type": "boolean", "optional": true, "description": "True if cookie is partitioned. Must be true if partitionKey is set." },
+                { "name": "partitionKey", "type": "string", "optional": true, "description": "Cookie partition key. If null and partitioned property is true, then key must be computed." },
```
we normally put optional properties last

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -604,6 +605,11 @@ static Ref<Inspector::Protocol::Page::Cookie> buildObjectForCookie(const Cookie&
         .setSecure(cookie.secure)
         .setSameSite(cookieSameSitePolicyJSON(cookie.sameSite))
         .release();
+    protocolCookie->setPartitioned(!cookie.partitionKey.isEmpty());
```
IMO this is unnecessary since we can always assume that `partitioned` is `true` if `partitionKey` is not empty

i think it's fine to leave `partitioned` in `Page.Cookie` since that's used in `Page.setCookie`, but i wouldn't provide it here since it's implicit based on `partitionKey` (and id add a comment in `Source/JavaScriptCore/inspector/protocol/Page.json` indicating this relationship)

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -695,6 +703,9 @@ static std::optional<Cookie> parseCookieObject(Inspector::Protocol::ErrorString&
     }
 
+    auto partitioned = cookieObject->getBoolean("partitioned"_s);
+    IsPartitionedCookie isPartitioned = partitioned && *partitioned ? IsPartitionedCookie::Yes : IsPartitionedCookie::No;
```
similarly, `partitionKey` should be the primary source of truth
```cpp
auto isPartitioned = (!partitionKey.isEmpty() || partitioned.value_or(false)) ? IsPartitionedCookie::Yes : IsPartitionedCookie::No;
```

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -753,7 +765,7 @@ Inspector::Protocol::ErrorStringOr<void> InspectorPageAgent::setCookie(Ref<JSON:
         if (!page)
             continue;
-        page->cookieJar().setRawCookie(*document, cookie.value());
+        page->cookieJar().setRawCookie(*document, cookie, isPartitioned);
```
wait actually is there really a need to include `partitioned` in `Page.Cookie` if it's only ever used by `Page.setCookie`?

what if instead we had `partitioned` just be another parameter for `Page.setCookie` since ultimately that's really only where it's needed (because in all other places we can assume `partitioned` if `partitionKey` is not empty)
```cpp
Inspector::Protocol::ErrorStringOr<void> InspectorPageAgent::setCookie(Ref<JSON::Object>&& cookieObject, std::optional<bool>&& shouldPartition)
{
    /* ... */
    auto isPartitioned = (!cookie->partitionKey.isEmpty() || shouldPartition.value_or(false)) ? IsPartitionedCookie::Yes : IsPartitionedCookie::No;
    /* ... */
        page->cookieJar().setRawCookie(*document, cookie.value(), isPartitioned);
    /* ... */
}
```
(and to be clear with this approach we'd only add `partitionKey` to `Page.Cookie`)

sysrqb replied:
Good point. That does simplify this a bit.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -224,6 +230,12 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
+        let partitionKeyInputRow = createInputRow("partitionKey", WI.UIString("Partition Key"), "text", data.partitionKey);
+        this._partitionKeyInputElement = partitionKeyInputRow.inputElement;
+        this._partitionKeyInputElement.readOnly = true;
+        partitionKeyInputRow.rowElement.hidden = !data.partitioned;
```
NIT: we dont need to do this here since we're gonna do it in `togglePartitionKeyRow`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -245,6 +257,18 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
         toggleExpiresRow();
 
+        let togglePartitionKeyRow = () => {
+            partitionKeyInputRow.rowElement.hidden = !this._partitionedCheckboxElement.checked;
```
we should probably also only show the `partitionKey` if it's set
```js
partitionKeyInputRow.rowElement.hidden = !this._partitionedCheckboxElement.checked || !data.partitionKey;
```
alternatively we could not even make the `partitionKeyInputRow` if the `partitionKey` is not set (e.g. when we're creating an entirely new cookie from scratch)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -224,6 +230,12 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
         this._secureCheckboxElement = createInputRow("secure", WI.unlocalizedString("Secure"), "checkbox", data.secure).inputElement;
 
+        this._partitionedCheckboxElement = createInputRow("partitioned", WI.unlocalizedString("Partitioned"), "checkbox", data.partitioned).inputElement;
+        let partitionKeyInputRow = createInputRow("partitionKey", WI.UIString("Partition Key"), "text", data.partitionKey);
```
what if instead of creating an entirely separate row we just showed the `<input readonly>` inline right after the `<input type="checkbox">`?  do you think developers need the "Partition Key" label in order to understand what they're looking at?  or is it obvious enough from the "Partitioned" label?

sysrqb replied:
I made this change. It introduces a slight glitch when the field is unhidden (at least in my testing), where the input field is initially shown on a separate line under the checkbox and then the layout is reflowed so the checkbox and field are on the same line. I'll need to dig into this a bit more, but it's good enough for now.

dcrousso replied:
gotcha.  if it's too complex i dont want this to get in the way of the overall feature, so im fine if you'd rather keep it as a separate row for now.  we can always come back after and improve the UI so long as the functionality is present :)

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -681,6 +686,8 @@ static std::optional<Cookie> parseCookieObject(Inspector::Protocol::ErrorString&
         return std::nullopt;
     }
 
+    cookie.partitionKey = cookieObject->getString("partitionKey"_s);
```
NIT: since this is optional we should probably do it last so that we don't do unnecessary work if something else fails

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -753,7 +760,8 @@ Inspector::Protocol::ErrorStringOr<void> InspectorPageAgent::setCookie(Ref<JSON:
         if (!page)
             continue;
-        page->cookieJar().setRawCookie(*document, cookie.value());
+        auto isPartitioned = shouldPartition.value_or(false) ? IsPartitionedCookie::Yes : IsPartitionedCookie::No;
```
please move this outside of the `for` since it doesn't rely on the `frame` at all

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -753,7 +760,8 @@ Inspector::Protocol::ErrorStringOr<void> InspectorPageAgent::setCookie(Ref<JSON:
         if (!page)
             continue;
-        page->cookieJar().setRawCookie(*document, cookie.value());
+        auto isPartitioned = shouldPartition.value_or(false) ? IsPartitionedCookie::Yes : IsPartitionedCookie::No;
```
if the developer modifies a cookie that has an existing `partitionKey` will this preserve that or generate a new one?  would be great to have some tests for this if possible :)

sysrqb replied:
It will recompute the key. I suppose this could be surprising behavior because the key that is shown in the popup could be different from the key that is actually set on the cookie. However, that should be impossible, because none of the values that can be modified in WebInspector would influence computed partition key, and we only show cookies in WebInspector that match the original partition key, so everything should match. An alternative is we could rename this enum from `IsPartitionedCookie` to `ShouldPartitionCookie`, and then we only set a new partition key if that is necessary.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Cookie.js:
```diff
@@ -25,7 +25,7 @@
 WI.Cookie = class Cookie
 {
-    constructor(type, name, value, {header, expires, session, maxAge, path, domain, secure, httpOnly, sameSite} = {})
+    constructor(type, name, value, {header, expires, session, maxAge, path, domain, secure, httpOnly, partitionKey, partitioned, sameSite} = {})
```
NIT: please match the ordering of properties from `Page.Cookie`

ditto below (i.e. `partitionKey` and `partitioned` should be after `sameSite`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -224,6 +237,16 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
+            partitionKeyElement.readOnly = true;
+            this._partitionKeyInputElement = partitionKeyElement;
+        }
+        this._partitionedCheckboxElement = createInputRow("partitioned", WI.unlocalizedString("Partitioned"), "checkbox", data.partitioned, this._partitionKeyInputElement).inputElement;
```
rather than have `additionalElement` is there not a way to do this in CSS?

as you have it now the `inputElement` returned by `createInputRow` will be the `<span>` which won't work for the logic in `createRow` or as efficiently with the add `addEventListener`

sysrqb replied:
I decided to go back to adding a dedicated row, and I'll improve it in a follow up.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -245,6 +268,20 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
 
+        if (this._partitionKeyInputElement) {
+            let togglePartitionKeyInput = () => {
+                this._partitionKeyInputElement.hidden = !this._partitionedCheckboxElement.childNodes[0].checked;
```
this is not ideal as i'd expect `this._partitionedCheckboxElement` to be equal to the checkbox, not some parent

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookieStorageContentView.js:
```diff
@@ -455,7 +465,8 @@ WI.CookieStorageContentView = class CookieStorageContentView extends WI.ContentV
             promises.push(target.PageAgent.deleteCookie(editingCookie.name, editingCookie.url));
-        promises.push(target.PageAgent.setCookie(cookieProtocolPayload));
+        let needsPartitionKey = !cookieToSet.partitionKey && !!cookieToSet.partitioned ? true : undefined;
+        promises.push(target.PageAgent.setCookie(cookieProtocolPayload, needsPartitionKey));
```
this will fail when inspecting older versions of iOS because `shouldPartition` doesn't exist
```js
// COMPATIBILITY (iOS 18.X, macOS 15.X): `Page.setCookie` did not have a `shouldPartition` parameter yet.
promises.push(target.PageAgent.setCookie.invoke({
    cookie: cookieProtocolPayload,
    shouldPartition: !cookieToSet.partitionKey && !!cookieToSet.partitioned,
}));
```

sysrqb replied:
I'll come back and fill in the `.X` later.

dcrousso replied:
FYI we normally do all of these at once when we snapshot a new protocol version

dcrousso reviewed (commented):
nice!  i think this is quite a bit nicer and simpler :)

thanks for iterating with me!  only a few remaining UI concerns

dcrousso commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -119,6 +120,19 @@ bool NetworkStorageSession::shouldBlockThirdPartyCookiesButKeepFirstPartyCookies
+        return;
+    }
+    auto partitionedCookie = cookie;
+    partitionedCookie.partitionKey = cookiePartitionIdentifier(firstParty);
```
should this also only happen if the `partitionKey` isn't set?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -224,6 +230,14 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
         this._secureCheckboxElement = createInputRow("secure", WI.unlocalizedString("Secure"), "checkbox", data.secure).inputElement;
 
+        this._partitionedCheckboxElement = createInputRow("partitioned", WI.unlocalizedString("Partitioned"), "checkbox", data.partitioned).inputElement;
+        let partitionKeyInputRow;
```
Style: we prefer to give all variables an initial value, so please add `= null`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
```diff
@@ -224,6 +230,14 @@ WI.CookiePopover = class CookiePopover extends WI.Popover
+        let partitionKeyInputRow;
+        if (data.partitionKey) {
+            partitionKeyInputRow = createInputRow("partition-key", WI.UIString("Partition Key"), "text", data.partitionKey);
+            this._partitionKeyInputElement = partitionKeyInputRow.inputElement;
```
we actually don't need to save this to a member variable since it's not used anywhere else, so id remove all `this._partitionKeyInputElement` and just have this be
```js
partitionKeyInputRow.inputElement.readOnly = true;
```

dcrousso reviewed (approved):
this looks great!  thanks so much for iterating with me :)

Commit:
Expose cookie Partition Key in Web Inspector
https://bugs.webkit.org/show_bug.cgi?id=279980
rdar://136293236

Reviewed by BJ Burg and Devin Rousso.

This patch adds a new column in the cookie storage view that displays a
cookie's partition key. Currently, the column is hidden by default. The
partition key is also shown and can also be set in the popover. However, the
values are not currently plumbed into the underlying cookie stores, so this
change doesn't have any affect on the underlying cookie.

* Source/JavaScriptCore/inspector/protocol/Page.json:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::buildObjectForCookie):
(WebCore::parseCookieObject):
(WebCore::InspectorPageAgent::setCookie):
* Source/WebCore/inspector/agents/InspectorPageAgent.h:
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::setRawCookie):
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setCookie):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/ShouldPartitionCookie.h: Added.
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/Cookie.js:
(WI.Cookie.parseSetCookieResponseHeader):
(WI.Cookie.prototype.get partitionKey):
(WI.Cookie.prototype.get partitioned):
(WI.Cookie.prototype.equals):
(WI.Cookie.prototype.toProtocol):
(WI.Cookie):
* Source/WebInspectorUI/UserInterface/Views/CookiePopover.js:
(WI.CookiePopover):
(WI.CookiePopover.prototype.get serializedData):
(WI.CookiePopover.prototype.show):
* Source/WebInspectorUI/UserInterface/Views/CookieStorageContentView.js:
(WI.CookieStorageContentView.prototype.initialLayout):
(WI.CookieStorageContentView.prototype._generateSortComparator):
(WI.CookieStorageContentView.prototype._showCookiePopover):
(WI.CookieStorageContentView.prototype.async _willDismissCookiePopover):
(WI.CookieStorageContentView.prototype._formatCookiePropertyForColumn):
(WI.CookieStorageContentView):
* Source/WebInspectorUI/UserInterface/Views/ResourceCookiesContentView.js:
(WI.ResourceCookiesContentView.prototype.tablePopulateCell):
(WI.ResourceCookiesContentView.prototype._refreshResponseCookiesSection):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setRawCookie):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::setRawCookie):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: https://commits.webkit.org/288103@main
